### PR TITLE
KAFKA-3863: System tests covering connector/task failure and restart

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
ping @Ishiihara @ewencp

File: connect/runtime/src/main/java/org/apache/kafka/connect/tools/MockConnector.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package org.apache.kafka.connect.tools;
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.common.utils.AppInfoParser;
+import org.apache.kafka.connect.connector.Connector;
+import org.apache.kafka.connect.connector.Task;
+import org.apache.kafka.connect.sink.SinkConnector;
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/tools/MockConnector.java]: Unused import failing checkstyle

File: connect/runtime/src/main/java/org/apache/kafka/connect/tools/MockConnector.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public List<Map<String, String>> taskConfigs(int maxTasks) {
+        return Collections.singletonList(config);
+    }
+
+    @Override
+    public void stop() {
+        if (executor != null)
+            executor.shutdownNow();
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/tools/MockConnector.java:95]: Given what runs in the executor, risk seems minimal, but shouldn't this include an `awaitShutdown`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/tools/MockSinkConnector.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.Task;
+import org.apache.kafka.connect.sink.SinkConnector;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Mock sink implementation which deletates to {@link MockConnector}.
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/tools/MockSinkConnector.java]: deletates -> delegates

File: tests/kafkatest/services/connect.py [hunk 4]
```diff
@@ -331,3 +337,46 @@ def start(self):
...
+            'connector.class': 'org.apache.kafka.connect.tools.MockSinkConnector',
+            'tasks.max': 1,
+            'topics': ",".join(self.topics),
+            'mock_mode': self.mode,
+            'delay_ms': self.delay_sec * 1000
+        })
+
+class MockSource(object):
+
+    def __init__(self, cc, topics, mode=None, delay_sec=10000, name="mock-source"):
```
Comment (ewencp) [inline tests/kafkatest/services/connect.py]: I think you forgot to reduce the default value here when you changed from ms to sec...

Comment (hachikuji) [inline tests/kafkatest/services/connect.py]: Yeah, good call. By the way, I'm investigating a build failure, so hold off merging for a little bit longer.

**Comment by hachikuji** [comment]
@ewencp Should be ready for another look. I'll run another branch build to be sure and link the results when it's done.

**Comment by hachikuji** [comment]
By the way, perhaps we should duplicate these tests to run against the standalone herder. The implementations are fairly different, so it ought to have some value.

**Comment by ewencp** [comment]
@hachikuji LGTM, let me know when the tests are done and I'll commit.

**Comment by ewencp** [comment]
@hachikuji Also, re: duplicating tests, I think the important question is which tests provide value and can be made to work with both easily. I'm not sure all the tests will be valuable. Even some some of the delivery semantics ones don't seem to have much value -- they really exercise the Worker*Task classes more than the herder.

**Comment by hachikuji** [comment]
@ewencp Needed one more fix, but looks like that was enough: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2016-06-22--001.1466572405--hachikuji--KAFKA-3863--7f4bf57/report.html. 

On duplicating tests, I agree that we should be selective. The tests in these patch are borderline I'd say. The mechanism for propagating state changes is different and should be tested, but you could argue that it's covered better in unit testing due to its simplicity.

**Comment by ewencp** [comment]
LGTM

**Pull request closed without merging**